rpc/dial.go

```diff
@@ -94,90 +101,137 @@ func dial(
 	tryLocal bool,
 ) (ClientConn, bool, error) {
 	var isJustDomain bool
-	if strings.HasPrefix(address, "unix://") {
+	switch {
+	case strings.HasPrefix(address, "unix://"):
 		dOpts.mdnsOptions.Disable = true
 		dOpts.webrtcOpts.Disable = true
 		dOpts.insecure = true
 		dOpts.disableDirect = false
-	} else {
-		if strings.ContainsRune(address, ':') {
-			isJustDomain = false
-		} else {
-			isJustDomain = net.ParseIP(address) == nil
-		}
+	case strings.ContainsRune(address, ':'):
+		isJustDomain = false
+	default:
+		isJustDomain = net.ParseIP(address) == nil
 	}
 
+	connCh := make(chan dialSuccess, 2)
+	errCh := make(chan error, 2)
+	var cancels []context.CancelFunc
 	if !dOpts.mdnsOptions.Disable && tryLocal && isJustDomain {
-		conn, cached, err := dialMulticastDNS(ctx, address, logger, dOpts)
-		if err != nil {
-			logger.Warnw("error dialing with mDNS; falling back to other methods", "error", err)
-		}
+		mdnsCtx, cancel := context.WithCancel(ctx)
+		cancels = append(cancels, cancel)
+		go func() {
+			conn, cached, err := dialMulticastDNS(mdnsCtx, address, logger, dOpts)
+			if err != nil {
+				logger.Warnw("error dialing with mDNS; falling back to other methods", "error", err)
+				errCh <- nil
+				return
+			}
 
-		if conn != nil {
-			return conn, cached, err
-		}
+			if conn != nil {
+				connCh <- dialSuccess{conn, cached, 0}
+				return
+			}
+
+			errCh <- nil
+		}()
 	}
 
 	if !dOpts.webrtcOpts.Disable {
-		signalingAddress := dOpts.webrtcOpts.SignalingServerAddress
-		if signalingAddress == "" || dOpts.webrtcOpts.AllowAutoDetectAuthOptions {
-			if signalingAddress == "" {
-				// try WebRTC at same address
-				signalingAddress = address
-			}
-			target, port, err := getWebRTCTargetFromAddressWithDefaults(signalingAddress)
-			if err != nil {
-				return nil, false, err
+		wrtcCtx, cancel := context.WithCancel(ctx)
+		cancels = append(cancels, cancel)
+		go func() {
+			signalingAddress := dOpts.webrtcOpts.SignalingServerAddress
+			if signalingAddress == "" || dOpts.webrtcOpts.AllowAutoDetectAuthOptions {
+				if signalingAddress == "" {
+					// try WebRTC at same address
+					signalingAddress = address
+				}
+				target, port, err := getWebRTCTargetFromAddressWithDefaults(signalingAddress)
+				if err != nil {
+					errCh <- err
+					return
+				}
+				fixupWebRTCOptions(dOpts, target, port)
+
+				// When connecting to an external signaler for WebRTC, we assume we can use the external auth's material.
+				// This path is also called by an mdns direct connection and ignores that case.
+				// This will skip all Authenticate/AuthenticateTo calls for the signaler.
+				if !dOpts.usingMDNS && dOpts.authMaterial == "" && dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial != "" {
+					logger.Debug("using signaling's external auth as auth material")
+					dOpts.authMaterial = dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial
+					dOpts.creds = Credentials{}
+				}
```

Comment:
TODO: `fixupWebRTCOptions` and the lines below mutate the dial options struct - need to either lock or pass copies since this operation could happen concurrently now.

Reply:
[done indirectly by copy dial options in the mDNS dial routine earlier](https://github.com/viamrobotics/goutils/pull/234/commits/9c86723c004e7af6626ed60517756ba530658d4d).

I really don't like how various stages of this dial pipeline modify options in place, but it might be too much to undo in this PR. I'm tempted to at least update all function signature to accept `dialOptions` values rather than pointers.